### PR TITLE
Add the ability to delete application

### LINKED ISSUE #1289: Allow deleting application
**What would you like to be added**:

Currently, we only allow disabling applications. The disabled application can be re-enabled anytime.
From the feedback of our users, they want a way to delete the application permanently.
So we should add a `deleted` flag into the `Application` model and provide an API to enable that feature.

**Why is this needed**:

### PR DESCRIPTION
**Pull request by nghialv**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1289

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/api/grpcapi/web_api.go [hunk 1]
```diff
@@ -481,6 +481,35 @@ func (a *WebAPI) DisableApplication(ctx context.Context, req *webservice.Disable
...
+	}
+
+	if err := a.validateAppBelongsToProject(ctx, req.ApplicationId, claims.Role.ProjectId); err != nil {
+		return nil, err
+	}
+
+	if err := a.applicationStore.DeleteApplication(ctx, req.ApplicationId); err != nil {
+		switch err {
+		case datastore.ErrNotFound:
+			return nil, status.Error(codes.InvalidArgument, "The application is not found")
```
Comment (khanhtc1202) [inline pkg/app/api/grpcapi/web_api.go:498]: should be `codes.NotFound` 👀

Comment (nghialv) [reply pkg/app/api/grpcapi/web_api.go]: Nice catch! Thanks.

**Comment by khanhtc1202** [comment]
nits, should we validate the application is deleted or not before update it?
ref: https://github.com/pipe-cd/pipe/blob/a1e9f8405a635052f36eb2eb514520cc55ff5302/pkg/datastore/applicationstore.go#L135

**Comment by nghialv** [comment]
@khanhtc1202 You are right. Added the missing validation.

File: pkg/datastore/applicationstore.go [hunk 2]
```diff
@@ -69,17 +71,33 @@ func (s *applicationStore) AddApplication(ctx context.Context, app *model.Applic
 func (s *applicationStore) EnableApplication(ctx context.Context, id string) error {
 	return s.ds.Update(ctx, applicationModelKind, id, applicationFactory, func(e interface{}) error {
 		app := e.(*model.Application)
+		if app.Deleted == true {
```
Comment (nakabonne) [inline pkg/datastore/applicationstore.go:74]: haha, this is what you said.

**Comment by nakabonne** [comment]
Thank you
/lgtm

File: pkg/datastore/applicationstore.go (hunk 2 above)
Comment (khanhtc1202) [reply pkg/datastore/applicationstore.go:74]: lol

**Comment by khanhtc1202** [comment]
I have another issue that needs opinions from you guys.
Currently, the `ListApplications` return all Application which belongs to selected Projects (validated accessibly before query)
https://github.com/pipe-cd/pipe/blob/96f3c68d722c0c1877f8b9a5e8bb50cd49f8a121/pkg/app/api/grpcapi/web_api.go#L601
I wonder, should we filter deleted application on the view (frontend) or we should filter it on web_api? How do you think about it 👀 @nakabonne @nghialv

File: pkg/datastore/applicationstore.go (hunk 2 above)
Comment (nghialv) [reply pkg/datastore/applicationstore.go:74]: oops. This is not what I said. lol. Let me fix this.
What I said is what @khanhtc1202 mentioned at https://github.com/pipe-cd/pipe/pull/1308#issuecomment-749355034

**Comment by nghialv** [comment]
@khanhtc1202 Yes. I prefer to filter the deleted ones at the server side (db level by updating the query). But currently, `firestore` does not provide a way to list all of the applications where the `deleted == true` || `deleted` field does not exist (old data).

**Comment by khanhtc1202** [comment]
> But currently, firestore does not provide a way to list all of the applications where the deleted == true || deleted field does not exist (old data).

I see, thanks for clarifying the situation 🙏 
Let's address that issue with another PR 👍

**Comment by nghialv** [comment]
Yes. That issue should be considered separately.
This PR is just for adding the RPC for @cakecatz.

**Comment by khanhtc1202** [comment]
Thanks 😊 
/approve

**Pull request merged**